stats.go

```diff
@@ -57,81 +57,244 @@ func (c *capabilities) Tagging() bool {
 }
 
 type counter struct {
-	prev        int64
-	curr        int64
-	cachedCount CachedCount
-}
-
-func newCounter(cachedCount CachedCount) *counter {
-	return &counter{cachedCount: cachedCount}
+	prev           int64
+	curr           int64
+	cachedCount    CachedCount
+	lastUpdateUnix int64
+	expired        uint32
+	scope          *scope
+	name           string
+}
+
+func newCounter(cachedCount CachedCount, name string, scope *scope) *counter {
+	return &counter{
+		cachedCount:    cachedCount,
+		scope:          scope,
+		name:           name,
+		lastUpdateUnix: globalNow().Unix(),
+	}
 }
 
 func (c *counter) Inc(v int64) {
 	atomic.AddInt64(&c.curr, v)
+
+	if atomic.LoadUint32(&c.expired) == 1 && c.scope != nil {
+		// Counter has expired, but direct ref was held and counter
+		// was incremented, therefore insert back into scope
+		scope := c.scope.getCurrentScope()
+
+		scope.cm.Lock()
+		if counters, ok := scope.counters[c.name]; !ok {
+			// No counters with the same name were created since expiry
+			scope.counters[c.name] = []*counter{c}
+		} else {
+			exists := false
+			for _, counter := range counters {
+				if counter == c {
+					// This counter was reinserted already by another routine
+					exists = true
+					break
+				}
+			}
+
+			if !exists {
+				// Another counter with the same name was created
+				// and this expired counter is not in the list so add to slice
+				scope.counters[c.name] = append(scope.counters[c.name], c)
+			}
+		}
+		scope.cm.Unlock()
+
+		atomic.StoreUint32(&c.expired, 0)
+	}
 }
 
 func (c *counter) value() int64 {
 	curr := atomic.LoadInt64(&c.curr)
-
 	prev := atomic.LoadInt64(&c.prev)
 	if prev == curr {
 		return 0
 	}
+
 	atomic.StoreInt64(&c.prev, curr)
 	return curr - prev
 }
 
-func (c *counter) report(name string, tags map[string]string, r StatsReporter) {
+func (c *counter) report(name string, tags map[string]string, r StatsReporter) bool {
 	delta := c.value()
-	if delta == 0 {
-		return
+	if delta != 0 {
+		r.ReportCounter(name, tags, delta)
+		atomic.StoreInt64(&c.lastUpdateUnix, globalNow().Unix())
+
+		return false
 	}
 
-	r.ReportCounter(name, tags, delta)
+	// Check if counter has expired
+	if c.scope != nil && globalNow().Unix() >
+		atomic.LoadInt64(&c.lastUpdateUnix)+c.scope.registry.expirePeriodSeconds {
+		atomic.StoreUint32(&c.expired, 1)
+		// Check once more for possible race where value updated between
+		// beginning of this call
+		delta = c.value()
```

Comment:
maybe move the logic of report + return if non-zero into an anonymous function to avoid duplicating the logic